packages/core/admin/_internal/node/vite/plugins.ts

```diff
@@ -0,0 +1,84 @@
+import path from 'node:path';
+import fs from 'node:fs';
+import { transformWithEsbuild, type Plugin } from 'vite';
+
+import { getDocumentHTML } from '../staticFiles';
+import type { BuildContext } from '../createBuildContext';
+
+const buildFilesPlugin = (ctx: BuildContext): Plugin => {
+  const CHUNK_ID = '.strapi/client/app.js';
+
+  return {
+    name: 'strapi/server/build-files',
+    apply: 'build',
+    buildStart() {
+      this.emitFile({
+        type: 'chunk',
+        id: CHUNK_ID,
+        name: 'strapi',
+      });
+    },
+    async generateBundle(_options, outputBundle) {
+      const bundle = outputBundle;
+      const entryFile = Object.values(bundle).find(
+        (file) =>
+          file.type === 'chunk' && file.name === 'strapi' && file.facadeModuleId?.endsWith(CHUNK_ID)
+      );
+
+      if (!entryFile) {
+        throw new Error(`Failed to find entry file in bundle (${CHUNK_ID})`);
+      }
+
+      if (entryFile.type !== 'chunk') {
+        throw new Error('Entry file is not a chunk');
+      }
+
+      const entryFileName = entryFile.fileName;
+      const entryPath = ['/admin'.replace(/\/+$/, ''), entryFileName].join('/');
+
+      this.emitFile({
+        type: 'asset',
+        fileName: 'index.html',
+        source: getDocumentHTML({
+          logger: ctx.logger,
+          props: {
+            entryPath,
+          },
+        }),
+      });
+    },
+  };
+};
+
+const runtimeRewritePlugin = (ctx: BuildContext): Plugin => {
+  return {
+    name: 'strapi/server/runtime-rewrite',
+    apply: 'serve',
+    resolveId(id) {
+      if (id === '/app.js') {
+        console.log('got the id', id, ctx.entry);
+        return ctx.entry;
+      }
+
+      return id;
+    },
+    configureServer(viteDevServer) {
+      return () => {
+        viteDevServer.middlewares.use(async (req, res, next) => {
+          console.log('req.url', req.url);
+          if (req.url === '/index.html') {
+            const filePath = path.relative(ctx.cwd, '.strapi/client/index.html');
+            const fileContents = fs.readFileSync(filePath, 'utf-8');
+            res.setHeader('Content-Type', 'text/html');
+            const html = await viteDevServer.transformIndexHtml(req.url, fileContents);
+            res.end(html);
+          }
```

Comment:
This serves the correct HTML file, although for some reason the browser gives a 404... The issue is I would have expected `transformIndexHtml` to see the `script` tag and bundle the code attached in dev mode....